domehunter/dome_control.py

```diff
@@ -0,0 +1,569 @@
+        Private method used for callback function of the encoder DOD.
+
+        Calling this method will toggle the encoder debug LED (if enabled)
+        and increment or decrement the encoder_count instance variable,
+        depending on the current rotation direction of the dome.
+
+        If the current dome direction cannot be determined, the last recorded
+        direction is adopted.
+        """
+        print(f"Encoder activated _increment_count")
+        self._change_led_state(1, leds=[LED_light.INPUT_1])
+
+        if self.current_direction.value is None:
+            self.current_direction = self.last_direction
+
+        if self.current_direction.name == "CW":
+            self.__encoder_count += 1
+        elif self.current_direction.name == "CCW":
+            self.__encoder_count -= 1
+        # TODO what is last_direction is also None?
+
+    def _az_to_ticks(self, az):
+        """
+        Convert degrees (azimuth) to equivalent in encoder tick count. Because
+        we zero the encoder at the home position, this conversion needs to
+        take into account the offset between the zero azimuth position and the
+        zero encoder position.
+
+        Parameters
+        ----------
+        az : astropy.coordinates.Longitude
+            Dome azimuth position in degrees.
+
+        Returns
+        -------
+        float
+            Returns encoder tick count corresponding to dome azimuth.
+
+        """
+        az_relative_to_home = (az - self.home_az).wrap_at(360 * u.degree)
+        return az_relative_to_home.degree / self.az_per_tick
+
+    def _ticks_to_az(self, ticks):
+        """
+        Convert encoder tick count to equivalent in degrees (azimuth).
+
+        Parameters
+        ----------
+        ticks : integer
+            The number of encoder ticks recorded.
+
+        Returns
+        -------
+        astropy.coordinates.longitude
+            The corresponding dome azimuth position in degrees.
+
+        """
+        tick_to_deg = Longitude(ticks * self.az_per_tick * u.deg)
+        return (self.home_az + tick_to_deg).wrap_at(360 * u.degree)
+
+    def _rotate_dome(self, direction):
+        """
+        Set dome to move clockwise.
+
+        Returns
+        -------
+        integer
+            Command status return code (tbd).
+
+        """
+        # if testing, deactivate the home_sernsor_pin to simulate leaving home
+        if self.testing and self.at_home:
+            self.__home_sensor_pin.drive_low()
+        # update the last_direction instance variable
+        self.last_direction = self.current_direction
+        # now update the current_direction variable to CW
+        self.current_direction = direction
+        # set the direction relay switch to CW position
+        if self.current_direction.name == "CW":
+            self.__direction_relay.on()
+            self._change_led_state(1, leds=[LED_light.RELAY_2_NO])
+            self._change_led_state(0, leds=[LED_light.RELAY_2_NC])
```

Comment:
Ideally all these LED operations should be done automatically by `when_activated` & `when_deactivated` callback functions associated with the relays. Unfortunately the gpiozero `DigitalOutputDevice` class doesn't have those attributes like `DigitalInput` does. But! We could fairly easily fix that oversight by making out own `DigitalOutputDevice` subclass that includes the `EventsMixin` though... I think it would be a simple as:

```
from gpiozero import DigitalInputDevice, DigitalOutputDevice
from gpiozero.mixins import EventsMixin

class EventfulDigitalOutputDevice(EventsMixin, DigitalOutputDevice):
    pass
```

Multiple inheritance is powerful magic...

Reply:
have given this a go and doesnt quite work, see issue #40